File: app/src/main/java/com/jpcn/chatapp/ui/fragments/ChatsFragment.java
package com.jpcn.chatapp.ui.fragments;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.firebase.messaging.FirebaseMessaging;
import com.jpcn.chatapp.R;
import com.jpcn.chatapp.model.ChatUser;
import com.jpcn.chatapp.model.Token;
import com.jpcn.chatapp.model.User;
import com.jpcn.chatapp.ui.adapters.UserAdapter;

import java.util.ArrayList;
import java.util.List;

public class ChatsFragment extends Fragment {

    private RecyclerView recyclerView;

    private UserAdapter userAdapter;
    private List<User> users;

    private List<ChatUser> chatUsers;

    FirebaseUser firebaseUser;
    DatabaseReference reference;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_chats, container, false);

        recyclerView = view.findViewById(R.id.recyclerViewChats);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        firebaseUser = FirebaseAuth.getInstance().getCurrentUser();

        chatUsers = new ArrayList<>();

        reference = FirebaseDatabase.getInstance().getReference("ChatUser").child(firebaseUser.getUid());
        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                chatUsers.clear();
                for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                    ChatUser chatList = snapshot.getValue(ChatUser.class);
                    chatUsers.add(chatList);
                }
                chatLists();
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });

        FirebaseMessaging.getInstance().getToken().addOnCompleteListener(new OnCompleteListener<String>() {
            @Override
            public void onComplete(@NonNull Task<String> task) {
                if (task.isSuccessful()) updateToken(task.getResult());
            }
        });

        return view;
    }

    private void updateToken(String valueToken) {
        DatabaseReference reference = FirebaseDatabase.getInstance().getReference("Tokens");
        Token token = new Token(valueToken);
        reference.child(firebaseUser.getUid()).setValue(token);
    }

    private void chatLists() {
        users = new ArrayList<>();

        reference = FirebaseDatabase.getInstance().getReference("users");
        reference.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                users.clear();
                for (DataSnapshot snapshot : dataSnapshot.getChildren()) {
                    User user = snapshot.getValue(User.class);
                    for (ChatUser chatUser : chatUsers) {
                        assert user != null;
                        if (user.getId().equals(chatUser.getId())) {
                            users.add(user);
                        }
                    }
                }
                userAdapter = new UserAdapter(getContext(), users, true);
                recyclerView.setAdapter(userAdapter);
            }

            @Override
            public void onCancelled(@NonNull DatabaseError databaseError) {

            }
        });
    }

}

File: app/src/main/java/com/jpcn/chatapp/ui/fragments/ChatsFragment.kt
package com.jpcn.chatapp.ui.fragments

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.messaging.FirebaseMessaging
import com.jpcn.chatapp.databinding.FragmentChatsBinding
import com.jpcn.chatapp.model.ChatUser
import com.jpcn.chatapp.model.Token
import com.jpcn.chatapp.model.User
import com.jpcn.chatapp.ui.adapters.UserAdapter

class ChatsFragment : Fragment() {
    lateinit var binding: FragmentChatsBinding
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View {
        binding = FragmentChatsBinding.inflate(inflater)
        binding.recyclerViewChats.setHasFixedSize(true)
        binding.recyclerViewChats.layoutManager = LinearLayoutManager(context)
        val firebaseUserId = FirebaseAuth.getInstance().currentUser?.uid.orEmpty()
        val chatUsers = mutableListOf<ChatUser>()
        FirebaseDatabase.getInstance().getReference("ChatUser").child(firebaseUserId)
            .addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    chatUsers.clear()
                    for (snapshot in dataSnapshot.children) {
                        snapshot.getValue(ChatUser::class.java)?.let {
                            chatUsers.add(it)
                        }
                    }
                    chatLists(chatUsers)
                }

                override fun onCancelled(databaseError: DatabaseError) {}
            })
        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
            if (task.isSuccessful) updateToken(task.result, firebaseUserId)
        }
        return binding.root
    }

    private fun updateToken(valueToken: String?, firebaseUserId: String) {
        val reference = FirebaseDatabase.getInstance().getReference("Tokens")
        val token = Token(valueToken)
        reference.child(firebaseUserId).setValue(token)
    }

    private fun chatLists(chatUsers: MutableList<ChatUser>) {
        val users = mutableListOf<User>()
        FirebaseDatabase.getInstance().getReference("users").addValueEventListener(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                users.clear()
                for (snapshot in dataSnapshot.children) {
                    val user = snapshot.getValue(User::class.java)
                    for (chatUser in chatUsers) {
                        assert(user != null)
                        if (user!!.id == chatUser.id) {
                            users.add(user)
                        }
                    }
                }
                val userAdapter = UserAdapter(context, users, true)
                binding.recyclerViewChats.adapter = userAdapter
            }

            override fun onCancelled(databaseError: DatabaseError) {}
        })
    }
}